grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -256,4 +253,67 @@ class PrepaymentInvoiceService {
+        Invoice invoice = Invoice.get(invoiceId)
+        if (!invoice) {
+            throw new IllegalArgumentException("Cannot find invoice with id: ${invoiceId}")
+        }
+        if (invoice.isPrepaymentInvoice || !invoice.hasPrepaymentInvoice) {
+            throw new IllegalArgumentException("Cannot update quantities on prepayment or non prepaid invoices")
+        }
+
+        items.each { Map item ->
+            updateInvoiceItemQuantity(item.id, item.quantity)
+        }
+    }
+
+    /**
+     * Updates invoice item's quantity and amount and finds related inverse item and updates quantity and amount
+     * */
+    void updateInvoiceItemQuantity(String itemId, Integer quantity) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (!invoiceItem) {
+            throw new IllegalArgumentException("Cannot find invoice with id: ${itemId}")
+        }
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("Quantity to change needs to be higher than 0")
+        }
+        if (!invoiceItem?.shipmentItem) {
+            throw new IllegalArgumentException("Invoice item is missing shipment item. Cannot edit quantity.")
+        }
+        Integer quantityAvailableToInvoice = invoiceItem.shipmentItem.quantityToInvoice
+        if (quantity > quantityAvailableToInvoice + invoiceItem.quantity) {
+            throw new IllegalArgumentException("Cannot update quantity to higher value than available to invoice")
+        }
+        // update invoice item's quantity (and adjust amount)
+        invoiceItem.quantity = quantity
+        invoiceItem.amount = quantity * invoiceItem.unitPrice
+        // update inverse item's quantity if there is inverse item (and adjust amount)
+        InvoiceItem inverseItem = findInverseItem(invoiceItem)
+        if (inverseItem) {
+            OrderItem orderItem = invoiceItem?.shipmentItem?.orderItems?.find { it }
+            InvoiceItem prepaymentItem = orderItem.invoiceItems.find { it.isPrepaymentInvoice }
+            // find quantity that is still available to inverse and add to the current quantity from this inverse item
+            // this needs to be checked because we can also increase invoiced quantity here
+            Integer quantityAvailableToInverse = inverseItem.quantity + getQuantityAvailableToInverse(orderItem, prepaymentItem)
+            Integer quantityToInverse = quantity > quantityAvailableToInverse ? quantityAvailableToInverse : quantity
+            inverseItem.quantity = quantityToInverse
+            inverseItem.amount = quantityToInverse * prepaymentItem.unitPrice * orderItem.order.prepaymentPercent * (-1)
+            return
+        }
+
+        // If there is a case that we did not found inverse item, but it was made available to inverse now after
+        // editing or removing invoice items on invoices, lets try to create it here
+        inverseItem = createInverseItemForShipmentItem(invoiceItem?.shipmentItem, invoiceItem)
```

Comment:
nitpick: I'd put this case in the if block (and change the condition to `if (!inverseItem)`) since it's much simpler and so the big block above doesn't need to be nested inside an if block.